zk_toolbox/crates/zk_inception/src/utils/ports.rs

```diff
@@ -0,0 +1,495 @@
+    }
+
+    pub fn allocate_port(&mut self, range: Range<u16>, info: String) -> anyhow::Result<u16> {
+        for port in range {
+            if !self.is_port_assigned(port) {
+                self.add_port_info(port, info.to_string());
+                return Ok(port);
+            }
+        }
+        anyhow::bail!(format!(
+            "No available ports in the given range. Failed to allocate port for: {}",
+            info
+        ));
+    }
+
+    pub fn allocate_ports_with_offset_from_defaults<T: ConfigWithChainPorts>(
+        &mut self,
+        config: &mut T,
+        chain_number: u32,
+    ) -> Result<()> {
+        let offset = ((chain_number - 1) as u16) * 100;
+        let port_range = (PORT_RANGE_START + offset)..PORT_RANGE_END;
+
+        let mut new_ports = HashMap::new();
+        for (desc, port) in T::get_default_ports() {
+            let mut new_port = port + offset;
+            if self.is_port_assigned(new_port) {
+                new_port = self.allocate_port(port_range.clone(), desc.clone())?;
+            } else {
+                self.add_port_info(new_port, desc.to_string());
+            }
+            new_ports.insert(desc, new_port);
+        }
+        config.set_ports(new_ports)?;
+        Ok(())
+    }
+}
+
+impl fmt::Display for EcosystemPorts {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let mut port_vec: Vec<_> = self.ports.iter().collect();
+        port_vec.sort_by_key(|&(key, _)| key);
+        for (port, port_infos) in port_vec {
+            for port_info in port_infos {
+                writeln!(f, "{} > {}", port_info, port)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Default for EcosystemPorts {
+    fn default() -> Self {
+        let mut ports = HashMap::new();
+        ports.insert(
+            DEFAULT_OBSERVABILITY_PORT,
+            vec!["Observability".to_string()],
+        );
+        Self {
+            ports: HashMap::new(),
+        }
+    }
+}
+
+pub struct EcosystemPortsScanner {}
+
+impl EcosystemPortsScanner {
+    /// Scans the ecosystem directory for YAML files and extracts port information.
+    /// Specifically, it looks for keys ending with "port" and collects their values.
+    /// Note: Port information from Docker Compose files will not be picked up by this method.
+    pub fn scan(shell: &Shell) -> Result<EcosystemPorts> {
+        let ecosystem_config = EcosystemConfig::from_file(shell)?;
+
+        // Create a list of directories to scan:
+        // - Ecosystem configs directory
+        // - Chain configs directories
+        // - ZKsync repo directory (docker-compose files)
+        let mut dirs = vec![ecosystem_config.config.clone()];
+        for chain in ecosystem_config.list_of_chains() {
+            if let Some(chain_config) = ecosystem_config.load_chain(Some(chain)) {
+                dirs.push(chain_config.configs.clone())
+            }
+        }
+        dirs.push(ecosystem_config.link_to_code);
```

Comment:
nit: actually, should it be `shell.current_dir()` instead?
if ecosystem is created outside of the zksync-era repo, then the actual docker compose file that is running will be inside ecosystem repo, right?

Reply:
nice catch!!